File: packages/wazeko-auth/src/qr.ts
import qrcodeTerminal from "qrcode-terminal";
import { QrCodeEvent } from "../../wazeko-types/src/index.js";

export class QrCodeManager {
  private currentQr: QrCodeEvent | null = null;

  updateQr(raw: string, attempts: number, timeoutSeconds: number = 60): QrCodeEvent {
    const event: QrCodeEvent = {
      raw,
      attempts,
      timeoutSeconds,
    };
    this.currentQr = event;
    return event;
  }

  printTerminal(small: boolean = true): void {
    if (this.currentQr) {
      console.log(`\n=== SCAN WHATSAPP QR CODE (Attempt #${this.currentQr.attempts}) ===`);
      qrcodeTerminal.generate(this.currentQr.raw, { small }, (qrcode) => {
        console.log(qrcode);
      });
      console.log("===================================================\n");
    }
  }

  get current(): QrCodeEvent | null {
    return this.currentQr;
  }
}

File: crates/wazeko-auth/src/qr.rs
use qrcode::render::unicode;
use qrcode::QrCode;
use tracing::info;
use wazeko_core::error::{Result, SessionError};
use wazeko_types::events::QrCodeEvent;

pub struct QrCodeManager {
    current_qr: Option<QrCodeEvent>,
}

impl Default for QrCodeManager {
    fn default() -> Self {
        Self::new()
    }
}

impl QrCodeManager {
    pub fn new() -> Self {
        Self { current_qr: None }
    }

    pub fn update_qr(&mut self, raw_qr: impl Into<String>, attempt: u32, timeout_seconds: u64) -> QrCodeEvent {
        let event = QrCodeEvent {
            raw: raw_qr.into(),
            attempts: attempt,
            timeout_seconds,
        };
        self.current_qr = Some(event.clone());
        event
    }

    pub fn render_terminal(raw_qr: &str) -> Result<String> {
        let code = QrCode::new(raw_qr.as_bytes())
            .map_err(|e| SessionError::KeyExchange(format!("Failed to generate QR code: {e}")))?;

        let image = code
            .render::<unicode::Dense1x2>()
            .dark_color(unicode::Dense1x2::Light)
            .light_color(unicode::Dense1x2::Dark)
            .build();

        Ok(image)
    }

    pub fn print_terminal(&self) -> Result<()> {
        if let Some(qr) = &self.current_qr {
            let rendered = Self::render_terminal(&qr.raw)?;
            info!(target: "wazeko::auth::qr", "\n{}", rendered);
            println!("\n=== SCAN WHATSAPP QR CODE (Attempt #{}) ===\n{}\n============================================\n", qr.attempts, rendered);
        }
        Ok(())
    }

    pub fn current(&self) -> Option<&QrCodeEvent> {
        self.current_qr.as_ref()
    }
}
